src/services/dsnp.ts

```diff
@@ -0,0 +1,267 @@
+ * events. The subscriptions start with all previous events which will be immediately emitted to
+ * the handlers, and then will continue listening for new events until the returned unsubscribe
+ * function is called.
+ * @param batchHandler a function called when a batch announcement is made on chain.
+ * @param registryHandler a function called when a DSNP registration is made or changed.
+ * @returns an unsubscribe function that will stop the subscriptions when called.
+ */
+export const startSubscriptions = async (
+  batchHandler: BatchAnnouncementHandler,
+  registryHandler: RegistryUpdateHandler
+): Promise<UnsubscribeFunction> => {
+  let blockNumber: number;
+  let blockIndex = 0;
+
+  // subscribe to all announcements
+  const unsubscribeToBatchPublications = await core.contracts.subscription.subscribeToBatchPublications(
+    (announcement: BatchPublicationLogData) => {
+      if (announcement.blockNumber !== blockNumber) {
+        blockNumber = announcement.blockNumber;
+        blockIndex = 0;
+      } else {
+        blockIndex++;
+      }
+      batchHandler(announcement, blockIndex);
+    },
+    { fromBlock: 1 }
+  );
+
+  // subscribe to registry events
+  const unsubscribeToRegistryUpdate = await core.contracts.subscription.subscribeToRegistryUpdates(
+    registryHandler,
+    { fromBlock: 1 }
+  );
+
+  return () => {
+    unsubscribeToRegistryUpdate();
+    unsubscribeToBatchPublications();
+  };
+};
+
+/**
+ * Load a DSNP batch file from a url in a batch announcment, parse it, and emit all rows in the batch
+ * to the given handler.
+ * @param batchAnnouncement batch announcment that specifies where to retrieve the batch file.
+ * @param rowHandler handler function that receives all rows from the batch.
+ * @throwss if batch cannot be read.
+ */
+export const readBatchFile = async (
+  batchAnnouncement: BatchPublicationLogData,
+  rowHandler: AnnouncementRowHandler
+): Promise<void> => {
+  let batchIndex = 0;
+  const reader = await core.batch.openURL(
+    (batchAnnouncement.fileUrl.toString() as any) as URL
+  );
+  await core.batch.readFile(reader, (announcementRow: AnnouncementType) =>
+    rowHandler(announcementRow, batchIndex++)
+  );
+};
+
+/**
+ * isGraphChangeAnnouncment returns whether the provided object is likely a GraphChangeAnnouncment
+ * @param obj any object
+ * @returns true if object appears to be a graph change announcment
+ */
+export const isGraphChangeAnnouncement = (
+  obj: unknown
+): obj is GraphChangeAnnouncement => {
+  return (
+    (obj as Record<string, unknown>)["announcementType"] ===
+    AnnouncementType.GraphChange
+  );
+};
+
+/**
+ * isBroadcastAnnouncment returns whether the provided object is likely a BroadcastAnnouncment
+ * @param obj any object
+ * @returns true if object appears to be a broadcast announcment
+ */
+export const isBroadcastAnnouncement = (
+  obj: unknown
+): obj is BroadcastAnnouncement => {
```

Comment:
There are separate validators for broadcast, reply and profile: https://libertydsnp.github.io/sdk-ts/modules/core_announcements_validation.html

Reply:
Yeah, the point of these is to segregate the content, not to validate it. There's been problems with validation in the past. Maybe we're past that now, but it would be nice if the SDK provided options for both.

Reply:
I think with v3 of the sdk we are past it, but if you just segregate, we shouldn't have the type narrowing or your later code that assumes the type can fail or be broken in odd ways.